### PR TITLE
ENH: add wkt / wkb ufuncs

### PR DESCRIPTION
**Pull request by jorisvandenbossche**
Start for adding from_wkt / to_wkt / from_wkb / to_wkt ufuncs.

They are right now already defined on the Geometry object, but those are useful as ufuncs as well. 
We can probably remove the implementation in pygeom.c to avoid duplication. But the question is whether we want to keep those as methods in the public API?

I should maybe also move this into a separate .c file, as the ufuncs.c is starting to become quite long? (although it are ufuncs)

Still need to make the nice Python functions, only the barebone ufuncs for now for the readers. 
Also still need to do proper error handling (ensure that I destroy the reader when there is an error)

### CODE REVIEW

**Comment by jorisvandenbossche** [comment]
Some other questions:

- Do we want to keep the hex / non-hex in a single ufunc? I suppose checking for each element individually which of the two it is will give some overhead. Also, you will typically either have all bytes or all hex, not a mixture?
- The hex input is also bytes, while you might typically have this as a string? (not sure myself) For example the `shapely.wkb.dumps(   hex=True)` gives a string.

File: src/ufuncs.c [hunk 1]
```diff
@@ -928,6 +929,211 @@ static void create_collection_func(char **args, npy_intp *dimensions,
...
+    /* Create the WKT writer */
+    writer = GEOSWKTWriter_create_r(context_handle);
+    if (writer == NULL) {
+        return;
+    }
+
+    GEOSWKTWriter_setRoundingPrecision_r(context_handle, writer, precision);
+    GEOSWKTWriter_setTrim_r(context_handle, writer, trim);
+    GEOSWKTWriter_setOutputDimension_r(context_handle, writer, dimension);
+    GEOSWKTWriter_setOld3D_r(context_handle, writer, use_old_3d);
```
Comment (caspervdw) [inline src/ufuncs.c]: I think these should be available as arguments to the functions. The only way I know to do that, is to actually add them to the ufunc inputs. So you'll get 5 input arguments. 

I do think they should e scalars, we don't want to call these 4 methods in the inner loop. I encountered the same problem in the buffer function, which I solved by asserting that it is a scalar.

File: src/ufuncs.c [hunk 2]
```diff
@@ -928,6 +929,211 @@ static void create_collection_func(char **args, npy_intp *dimensions,
...
+        PyObject **out = (PyObject **)op1;
+
+        if (in1 == NULL) {  
+            Py_XDECREF(*out);
+            Py_INCREF(Py_None);
+            *out = Py_None;
+        } else {
+            wkt = GEOSWKTWriter_write_r(context_handle, writer, in1);
+            Py_XDECREF(*out);
+            *out = PyUnicode_FromFormat(format, wkt);
```
Comment (caspervdw) [inline src/ufuncs.c]: In this case `PyUnicode_FromString` is better I think

File: src/ufuncs.c [hunk 3]
```diff
@@ -928,6 +929,211 @@ static void create_collection_func(char **args, npy_intp *dimensions,
...
+    reader = GEOSWKBReader_create_r(context_handle);
+    if (reader == NULL) {
+        return;
+    }
+
+    UNARY_LOOP {
+        in1 = *(PyObject **)ip1;
+
+        /* Cast the PyObject (only bytes) to char* */
+        if (!PyBytes_Check(in1)) {
```
Comment (caspervdw) [inline src/ufuncs.c]: I guess we also need to accept `Py_None`

File: src/ufuncs.c [hunk 4]
```diff
@@ -928,6 +929,211 @@ static void create_collection_func(char **args, npy_intp *dimensions,
...
+        if (wkb == NULL) {
+            return;
+        }
+
+        /* Check if this is a HEX WKB */
+        if (size != 0) {
+            is_hex = ((wkb[0] == 48) | (wkb[0] == 49));
+        } else {
+            is_hex = 0;
+        }
```
Comment (caspervdw) [inline src/ufuncs.c]: I think the performance hit here is negligible, so keep it imho

File: src/ufuncs.c [hunk 5]
```diff
@@ -928,6 +929,211 @@ static void create_collection_func(char **args, npy_intp *dimensions,
...
+
+        /* Cast the PyObject (bytes or str) to char* */
+        if (PyBytes_Check(in1)) {
+            wkt = PyBytes_AsString(in1);
+            if (wkt == NULL) { return; }
+        }
+        else if (PyUnicode_Check(in1)) {
+            wkt = PyUnicode_AsUTF8(in1);
+            if (wkt == NULL) { return; }
+        } else {
```
Comment (caspervdw) [inline src/ufuncs.c]: Same comment as above, we should also accept `None`

File: src/ufuncs.c [hunk 6]
```diff
@@ -928,6 +929,211 @@ static void create_collection_func(char **args, npy_intp *dimensions,
...
+                        npy_intp *steps, void *data)
+{
+    void *context_handle = geos_context[0];
+    GEOSGeometry *in1;
+
+    GEOSWKBWriter *writer;
+    unsigned char *wkb;
+    size_t size;
+    
+    int dimension = 3;
```
Comment (caspervdw) [inline src/ufuncs.c]: See comment in the to_wkt function, I think these should be arguments

File: pygeos/test/test_readers_writers.py [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+
+from pygeos.ufuncs import from_wkt, from_wkb, to_wkt, to_wkb
+
+
+POINT11_WKB = (
+    b"\x01\x01\x00\x00\x00\x00\x00\x00\x00\x00\x00\xf0?\x00\x00\x00\x00\x00\x00\xf0?"
+)
+
+
+def test_from_wkt():
```
Comment (caspervdw) [inline pygeos/test/test_readers_writers.py]: There is some of overlap with the wkb/wkt tests in `test_geometry`, ideally these should be merged

**Comment by caspervdw** [comment]
Thanks for this addition! I think the wkt/wkb functions on the Geometry object should either be dropped, or should make use of the ufuncs. I don't like the duplicate code.

I guess the hex input might also be a string indeed. Maybe best to try to parse both bytes and strings in the ufunc.

File: src/ufuncs.c [hunk 8]
```diff
@@ -928,6 +929,211 @@ static void create_collection_func(char **args, npy_intp *dimensions,
...
+    char is_hex;
+
+    /* Create the WKB reader */
+    reader = GEOSWKBReader_create_r(context_handle);
+    if (reader == NULL) {
+        return;
+    }
+
+    UNARY_LOOP {
+        in1 = *(PyObject **)ip1;
```
Comment (jorisvandenbossche) [inline src/ufuncs.c]: @caspervdw can you explain this line? Why the double `**` ?

Getting this right was what I first didn't get to work, but in the end found this in one of the other functions you already wrote with object dtype input.

File: src/ufuncs.c (hunk 1 above)
Comment (jorisvandenbossche) [reply src/ufuncs.c]: That means that you can't give them default values? (on the ufunc level, can still do that in the python wrapper)

**Comment by jorisvandenbossche** [comment]
Thanks for the comments! 
I already addressed them partly. For `to_wkt`, I already added the parameters. But, that brings up a new question: what should be the defaults? 
For now, I used the defaults as they are in GEOS. But, those are not necessarily the most user friendly defaults (shapely also uses `trim=True` and `output_dimension=3` as default, which deviates from GEOS).

File: src/ufuncs.c (hunk 8 above)
Comment (caspervdw) [reply src/ufuncs.c]: Yes, I struggled with that too. The `(Pyobject **)` tells the interpreter that `ip1` is a pointer to a `PyObject *`. 

Then we take the value of that (with the extra star at the start), so that `in1` becomes a `PyObject *`.

**Comment by caspervdw** [comment]
I agree that it is better here to deviate from the GEOS defaults. In the `ToWKB` I wrote earlier I used:

```
char trim = 1;
int precision = 6;
int dimension = 3;
int use_old_3d = 0;
```

**Comment by caspervdw** [comment]
The ToWKT part is good to go. I like the docstring!

**Comment by jorisvandenbossche** [comment]
So I will:

- remove the methods on the Geometry object (and see how to use the to_wkt ufunc in the repr of that object, this might be a hard one)
- convert the existing tests for those to use the ufuncs / consolidate it with my tests
- also add python wrappers for the from_wkt/from_wkb/to_wkb functions

That sounds good?

**Comment by jorisvandenbossche** [comment]
For the hex/bytes question, I will add accepting hex-strings instead of only bytes. However, for the actual bytes representation, I would personally only accept bytes, and not strings to be interpreted as bytes (that feels cleaner).

> I think the performance hit here is negligible, so keep it imho

Yes, it's indeed only a minor overhead (2-3% when I checked, but that is also in the uncertainty range). 
A reason to still have it as a keyword instead of automatic detection, is that I would find it very strange to have a mixture of bytes / hex in a single array. I am not sure we should support that.

---

BTW, I did some performance timings in general, and this PR gives exactly the same performance (as expected) as the cython version I wrote before, and it gives around a 20x speed-up for `from_wkb`/`to_wkb` and a 10x speed-up for `from_wkt`/`to_wkt` (for an array of Points, so which is the "easiest" (where the looping is relatively more overhead compared to the actual GEOS call), with bigger geoms the speed-up will be a bit less).

**Comment by caspervdw** [comment]
> However, for the actual bytes representation, I would personally only accept bytes, and not strings to be interpreted as bytes (that feels cleaner).

I agree, WKB should always be `bytes` and never `string`. `WKT` could be both.

> Yes, it's indeed only a minor overhead (2-3% when I checked, but that is also in the uncertainty range). I would find it very strange to have a mixture of bytes / hex in a single array. I am not sure we should support that.

The mixture of both would be odd indeed, but that's no reason to not implement it in this way, right? Maybe adding it as a keyword argument (`hex=True/False`) like in the `ToWKB` is the way to go then. I would be happy with both options.

**Comment by jorisvandenbossche** [comment]
OK, made a bunch of updates:

- added python wrappers for the other methods as well
- added parameters to `to_wkb`
- consolidated the tests (and added a `str(Geometry)` functionality to the GeometryObject to return default WKT)
- return hex as string in `to_wkb`, and accept strings in `from_wkb`

I already updated the tests to no longer use the `Geometry.to/from_wkb/wkt` methods, but did no yet finally remove the methods, because I am running in to the following issue where `Geometry.from_wkb` works, but the ufunc version `pygeos.from_wkb` does not (so the tests that I converted to use the ufuncs are failing):

```
In [1]: import pygeos

In [2]: p = pygeos.points(1, 1)

In [3]: p.to_wkb() == pygeos.to_wkb(p) 
Out[3]: True

In [4]: p.to_wkb(byte_order=0) == pygeos.to_wkb(p, byte_order=0)
Out[4]: True

In [5]: wkb = pygeos.to_wkb(p, byte_order=0)

In [6]: pygeos.Geometry.from_wkb(wkb) 
Out[6]: <pygeos.Geometry POINT (1 1)>

In [7]: pygeos.from_wkb(wkb) 
---------------------------------------------------------------------------
GEOSException                             Traceback (most recent call last)
<ipython-input-7-7d2b49c9458f> in <module>
----> 1 pygeos.from_wkb(wkb)

~/scipy/repos/pygeos/pygeos/io.py in from_wkb(geometry, **kwargs)
    172 
    173     """
--> 174     return lib.from_wkb(geometry, **kwargs)

GEOSException: ParseException: Unexpected EOF parsing WKB
```

This example is with a different byte order, but it also occurs with the default byte order for a different geometry type, eg:

```
In [8]: p = pygeos.Geometry("POLYGON ((0 0, 2 0, 2 2, 0 2, 0 0))") 

In [9]: p.to_wkb() == pygeos.to_wkb(p) 
Out[9]: True

In [10]: wkb = pygeos.to_wkb(p)

In [11]: pygeos.Geometry.from_wkb(wkb) 
Out[11]: <pygeos.Geometry POLYGON ((0 0, 2 0, 2 2, 0 2, 0 0))>

In [12]: pygeos.from_wkb(wkb) 
---------------------------------------------------------------------------
GEOSException                             Traceback (most recent call last)
<ipython-input-12-7d2b49c9458f> in <module>
----> 1 pygeos.from_wkb(wkb)

~/scipy/repos/pygeos/pygeos/io.py in from_wkb(geometry, **kwargs)
    172 
    173     """
--> 174     return lib.from_wkb(geometry, **kwargs)

GEOSException: ParseException: Unexpected EOF parsing WKB
```

but I don't directly see the difference between the code of the ufunc or the object's method ..

**Comment by jorisvandenbossche** [comment]
One additional question regarding naming: currently I am using the `to_wkt` / `from_wkt` pattern. But, it could also be `as_wkt` / `from_wkt` (more similar as the postgis terminology) or `read_wkt` / `write_wkt` (more close to GEOS terminology of WKTReader/Writer)

**Comment by jorisvandenbossche** [comment]
OK, looked into the failure I described above (difference between `ufuncs::from_wkb` and `Geometry.from_wkb`), by printing the wkb chars passed to the GEOS function in c:

```
In [1]: import pygeos 
   ...: p = pygeos.Geometry("POLYGON ((0 0, 2 0, 2 2, 0 2, 0 0))")  
   ...: wkb = pygeos.to_wkb(p)

In [2]: pygeos.from_wkb(wkb) 
Size: 77
01 03 00 00 00 01 00 00 00 05 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
00 00 00 00 00 40 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 40 00 00 00 00 00 00 00 40 00 
00 00 00 00 00 00 00 00 00 00 00 00 00 00 40 
---------------------------------------------------------------------------
GEOSException                             Traceback (most recent call last)
...
GEOSException: ParseException: Unexpected EOF parsing WKB

In [3]: pygeos.Geometry.from_wkb(wkb)
Size: 93
01 03 00 00 00 01 00 00 00 05 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
00 00 00 00 00 40 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 40 00 00 00 00 00 00 00 40 00 
00 00 00 00 00 00 00 00 00 00 00 00 00 00 40 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
Out[3]: <pygeos.Geometry POLYGON ((0 0, 2 0, 2 2, 0 2, 0 0))>
```

So you can see that although the same python object is passed, in the working case there are additional null bytes, while those have been removed in the failing ufunc case. 
And the original python object has indeed those null bytes:

```
In [7]: wkb
Out[7]: b'\x01\x03\x00\x00\x00\x01\x00\x00\x00\x05\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00@\x00\x00\x00\x00\x00\x00\x00
\x00\x00\x00\x00\x00\x00\x00\x00@\x00\x00\x00\x00\x00\x00\x00@\x00\x00\x00\x00\x00\x00\x00
\x00\x00\x00\x00\x00\x00\x00\x00@\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00
\x00\x00'

In [8]: len(wkb)
Out[8]: 93
```

And what is more, this only happens for the ufunc if you pass a scalar, not if you actually pass an array:

```
In [6]: pygeos.from_wkb(np.array([wkb], dtype=object))
Size: 93
01 03 00 00 00 01 00 00 00 05 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 40 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 40 00 00 00 00 00 00 00 40 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 40 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
Out[6]: 
array([<pygeos.Geometry POLYGON ((0 0, 2 0, 2 2, 0 2, 0 0))>],
      dtype=object)
```

So there seems to be going something wrong how numpy treats a non-numpy array as input to the ufunc in this case. I don't know if it is something that I am doing wrong, or if it is a bug in numpy. 
But at least it provides a way to workaround it in this case.

**Comment by jorisvandenbossche** [comment]
> I don't know if it is something that I am doing wrong, or if it is a bug in numpy.
But at least it provides a way to workaround it in this case.

Ah, this is because of the default dtype inference of numpy. If you pass a bytes like object to an array (without specifying a dtype), you get a `S` dtyped array, and accessing a single element of that removes trailing null bytes:

```
In [24]: np.array(wkb) 
Out[24]: 
array(b'\x01\x03\x00\x00\x00\x01\x00\x00\x00\x05\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00
\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00@\x00\x00\x00\x00\x00\x00\x00
\x00\x00\x00\x00\x00\x00\x00\x00@\x00\x00\x00\x00\x00\x00\x00@\x00\x00\x00\x00\x00\x00\x00
\x00\x00\x00\x00\x00\x00\x00\x00@',
      dtype='|S93')

In [25]: np.array(wkb).item() 
Out[25]: b'\x01\x03\x00\x00\x00\x01\x00\x00\x00\x05\x00\x00\x00\x00\x00\x00\x00\x00\x00
\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00@\x00\x00\x00\x00\x00
\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00@\x00\x00\x00\x00\x00\x00\x00@\x00\x00\x00\x00\x00
\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00@'

In [26]: wkb 
Out[26]: b'\x01\x03\x00\x00\x00\x01\x00\x00\x00\x05\x00\x00\x00\x00\x00\x00\x00\x00\x00
\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00@\x00\x00\x00\x00\x00
\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00@\x00\x00\x00\x00\x00\x00\x00@\x00\x00\x00\x00\x00
\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00@\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00'
```

So a roundtrip to numpy `S93` and back to object removes the trailing null bytes.  
Some explanation in https://stackoverflow.com/questions/37435309/numpy-trimming-trailing-zeros-in-byte-strings and https://github.com/numpy/numpy/issues/3878. The reason they do that because the trailing null bytes as end-of-string is the only way to have shorter strings in a fixed-length string dtype, but for our use case this is of course totally unacceptable. So we probably should disallow those string dtypes as input.

**Comment by jorisvandenbossche** [comment]
@caspervdw the ufuncs are now fully working, but a remaining item: can we deduplicate the code between ufuncs.c and pygeom.c ? Because pygeom.c does still need to have basic read/write functionality for the repr and the constructor (but without options). 
But I don't directly see an easy way to reuse the function we defined for the ufunc. You would need to define the arguments in the way the the ufunc function expects them, I don't directly succeed in that.

File: pygeos/io.py [hunk 9]
```diff
@@ -0,0 +1,178 @@
...
+    Specification for SQL <https://www.opengeospatial.org/standards/sfs>`__.
+
+    Parameters
+    ----------
+    geometry : Geometry or array_like
+    rounding_precision : int, default 6
+        The rounding precision when writing the WKT string. Set to a value of
+        -1 to indicate the full precision.
+    trim : bool, default True
+        Whether to trim unnecessary decimals (trailing zero's).
```
Comment (caspervdw) [inline pygeos/io.py]:
```suggestion
        Whether to trim unnecessary decimals (trailing zeros).
```

File: pygeos/io.py [hunk 10]
```diff
@@ -0,0 +1,178 @@
...
+
+    Examples
+    --------
+    >>> to_wkt(Geometry("POINT (0 0)"))
+    'POINT (0 0)'
+    >>> to_wkt(Geometry("POINT (0 0)"), rounding_precision=3, trim=False)
+    'POINT (0.000 0.000)'
+    >>> to_wkt(Geometry("POINT (0 0)"), rounding_precision=-1, trim=False)
+    'POINT (0.0000000000000000 0.0000000000000000)'
+
```
Comment (caspervdw) [inline pygeos/io.py]: Can you remove the empty line?

File: pygeos/io.py [hunk 11]
```diff
@@ -0,0 +1,178 @@
...
+        Whether the SRID should be included in WKB (this is an extension
+        to the OGC WKB specification).
+
+    Examples
+    --------
+    >>> to_wkb(Geometry("POINT (1 1)"))
+    b'\x01\x01\x00\x00\x00\x00\x00\x00\x00\x00\x00\xf0?\x00\x00\x00\x00\x00\x00\xf0?'
+    >>> to_wkb(Geometry("POINT (1 1)"), hex=True)
+    '0101000000000000000000F03F000000000000F03F'
+
```
Comment (caspervdw) [inline pygeos/io.py]: Can you remove the empty line?

File: pygeos/io.py [hunk 12]
```diff
@@ -0,0 +1,178 @@
...
+    Parameters
+    ----------
+    geometry : str or array_like
+        The WKB byte object(s) to convert.
+
+    Examples
+    --------
+    >>> from_wkb(b'\x01\x01\x00\x00\x00\x00\x00\x00\x00\x00\x00\xf0?\x00\x00\x00\x00\x00\x00\xf0?')
+    <pygeos.Geometry POINT (1 1)>
+
```
Comment (caspervdw) [inline pygeos/io.py]: Can you remove the empty line?

File: pygeos/test/test_io.py [hunk 13]
```diff
@@ -0,0 +1,165 @@
...
+def test_from_wkt_all_types(geom):
+    wkt = pygeos.to_wkt(geom)
+    actual = pygeos.from_wkt(wkt)
+    assert pygeos.equals(actual, geom)
+
+
+@pytest.mark.parametrize(
+    "wkt", ("POINT EMPTY", "LINESTRING EMPTY", "GEOMETRYCOLLECTION EMPTY")
+)
+def test_from_wkt_empty(wkt):
```
Comment (caspervdw) [inline pygeos/test/test_io.py]: Can you also make this test for the WKB?  I vaguely remember segfaults when trying to serialize and deserialize empty geometries from an to WKB (especially the POINT EMPTY)

**Comment by jorisvandenbossche** [comment]
> I think we should investigate the overhead if we define part of the Geometry object in pure python. Maybe create the base class in C (with the static _ptr attribute), and subclass it in Python adding extra features such as the to_wkb.

But the ufuncs would still return the base extension type. So do you mean that in the python functions, we would (for those methods that return geometries) loop through it and wrap each element in the python class?

**Comment by caspervdw** [comment]
Maybe we could make a kind of accessor like:

Geometry.obj

That creates a Python object with all the bells and whistles

**Comment by jorisvandenbossche** [comment]
Will update for your comments when having some time this week.

One more question about the naming: I now used `to_wkt` / `from_wkt` pattern. But, it could also be `write_wkt` / `read_wkt` (in line of GEOS's Writer/Reader classes) or `as_wkt` / `to_wkt` (more like postgis).

**Comment by caspervdw** [comment]
I don't like the "read" and "write" namings, as we also do not speak of "readers" and "writers" (that is a bit C-ish). We just convert representations of geometries. 

I like the idea of aligning with PostGIS `as_wkt` / `from_wkt`, but Python packages often use the `to_` and `from_` prefixes (e.g. `pandas.DataFrame.to_csv` or `numpy.ndarray.to_list`)

So my preference is keeping it as is. Also the least amount of work.

**Comment by jorisvandenbossche** [comment]
One more question: are you OK with dropping WKB input in the `Geometry(..)` constructor? 
For WKT it is convenient to be able to do `Geometry("POINT (1 1)"`), but for WKB I think people can just use `from_wkb` ? (then only the WKT io has some duplication in pygeom.c)

**Comment by jorisvandenbossche** [comment]
- cleaned up empty lines
- splitted tests (not to a single assert per test, but at least more splitted)
- added test for empty geometries (empty point does not work for WKB, as expected, but does still give a wrong error (MemoryError), so that is still a TODO)
- removed the methods on the Geometry object (and simplified the WKT code, since that doesn't need to handle with options any more -> that is covered by pygeos.to/from_wkt)

**Comment by jorisvandenbossche** [comment]
And the "POINT EMPTY" error in to_wkb is now also fixed: it still raises an error, but now the proper one with a informative error message:

```
In [2]: geom = pygeos.Geometry("POINT EMPTY") 

In [3]: pygeos.to_wkb(geom)
---------------------------------------------------------------------------
GEOSException                             Traceback (most recent call last)
<ipython-input-3-2272ebf49184> in <module>
----> 1 pygeos.to_wkb(geom)

~/scipy/repos/pygeos/pygeos/io.py in to_wkb(geometry, hex, output_dimension, byte_order, include_srid, **kwargs)
    127         np.intc(byte_order),
    128         np.bool(include_srid),
--> 129         **kwargs,
    130     )
    131 

GEOSException: IllegalArgumentException: Empty Points cannot be represented in WKB

```
